src/tree-router.js

```diff
@@ -0,0 +1,287 @@
+   */
+  constructor (routeRegistry) {
+    this._routeRegistry = routeRegistry;
+  }
+
+  /**
+   * implementation of catbee args provider interface
+   * @param {URI} location
+   * @returns {Object} signal and args
+   */
+  getArgsAndSignalByUri (location) {
+    let route = this._routeRegistry.matchingRoute(location.path);
+    return route ? route.handle(location) : {};
+  }
+}
+
+
+/**
+ * Service for getting route information in actions
+ */
+class TreeRouterService {
+  /**
+   * Creates router service using supplied registry
+   * @param {RouteRegistry} routeRegistry
+   */
+  constructor (routeRegistry) {
+    this._routeRegistry = routeRegistry;
+  }
+
+  /**
+   * Returns path of a parent route substituting path and query params for provided args
+   * @param {String} name of a current route
+   * @param {Object} args used for path and query params substitution
+   * @returns {String} path of a  parent route
+   */
+  parent (name, args) {
+    let current = this._routeRegistry.byName(name);
+    let parent = current ? current.parent : null;
+    return parent ? this._replaceParams(parent.path(), args) : null;
+  }
+
+  /**
+   * Returns path of a named route substituting path and query params for provided args
+   * @param {String} name of a route to get
+   * @param {Object} args used for path and query params substitution
+   * @returns {String} path of a named route
+   */
+  named (name, args) {
+    let found = this._routeRegistry.byName(name);
+    if (!found) {
+      return null;
+    }
+    return this._replaceParams(found.path(), args);
+  }
+
+  _replaceParams (path, args) {
+    return Object.keys(args || {}).reduce(
+      (acc, curr) => acc.replace(':' + curr, args[curr]), path);
+  }
+}
+
+
+/**
+ * Route item, encapsulates tree hiearachy, stores handlers for each route
+ */
+class Route {
+  /**
+   * Creates new route instance
+   * @param {String} url - route url, will be used instead of name if latter is falsy
+   * @param {String} name - route name, can be used to identify route
+   * @param {Object} handler - route handler (signal + args)
+   */
+  constructor (url, name, handler) {
+    this._url = url.startsWith('/') ? url : '/' + url;
+    this._name = name;
+    this._handler = handler;
+    this._parent = null;
+  }
+
+  _compiledPath () {
+    return routeHelper.compileRoute(new URI(this.path()));
+  }
+
+  /**
+   * Returns true if route can handle specified path
+   * @param {String} path
+   * @returns {Boolean}
+   */
+  matches (path) {
+    let compiled = this._compiledPath();
+    return compiled.expression.test(path);
+  }
+
+  /**
+   * Returns signal and args that handles current route
+   * @param {String} uri of a handled request
+   * @returns {Object} handler object (signal + args)
+   */
+  handle (uri) {
+    let compiled = this._compiledPath();
+    let uriArgs = compiled.map(new URI(uri));
+    let args = Object.assign({ routeName: this.name }, uriArgs, this._handler.args);
```

Comment:
I really have strong doubts about magical `routeName` argument. Perhaps it should be at least prefixed with the `$` sign, or better we could think about another way of passing route name.